source/extensions/filters/http/rbac/rbac_filter.h

```diff
@@ -43,20 +44,21 @@ class RoleBasedAccessControlFilterConfig {
 
   Filters::Common::RBAC::RoleBasedAccessControlFilterStats& stats() { return stats_; }
 
-  const absl::optional<Filters::Common::RBAC::RoleBasedAccessControlEngineImpl>&
+  const Filters::Common::RBAC::RoleBasedAccessControlEngineImpl*
   engine(const Router::RouteConstSharedPtr route,
          Filters::Common::RBAC::EnforcementMode mode) const;
 
 private:
-  const absl::optional<Filters::Common::RBAC::RoleBasedAccessControlEngineImpl>&
+  const Filters::Common::RBAC::RoleBasedAccessControlEngineImpl*
   engine(Filters::Common::RBAC::EnforcementMode mode) const {
-    return mode == Filters::Common::RBAC::EnforcementMode::Enforced ? engine_ : shadow_engine_;
+    return mode == Filters::Common::RBAC::EnforcementMode::Enforced ? engine_.get()
+                                                                    : shadow_engine_.get();
   }
 
   Filters::Common::RBAC::RoleBasedAccessControlFilterStats stats_;
 
-  const absl::optional<Filters::Common::RBAC::RoleBasedAccessControlEngineImpl> engine_;
-  const absl::optional<Filters::Common::RBAC::RoleBasedAccessControlEngineImpl> shadow_engine_;
+  std::unique_ptr<Filters::Common::RBAC::RoleBasedAccessControlEngineImpl> engine_;
```

Comment:
why did you need this change, because of NonCopyable? I thought optional can hold NonCopyable class, no?

Reply:
I think so. Optional is a value type, I believe. The factory call is what gave me the compilation issue.

Reply:
ok no big deal, make it std::unique_ptr<const ...>?

Reply:
Optional stores its content inline, so yes, it requires a copy if you pass an optional from a factory. The other way would be to store a reference in the optional, but then someone needs to own the engine anyways.

Reply:
> Optional stores its content inline, so yes, it requires a copy if you pass an optional from a factory. The other way would be to store a reference in the optional, but then someone needs to own the engine anyways.

I thought that can be addressed by std::move, it is fine anyway.